subsys/bluetooth/mesh/gatt_cli.c

```diff
@@ -0,0 +1,337 @@
+/*
+ * Copyright (c) 2021 Xiaomi Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/byteorder.h>
+
+#include <net/buf.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/mesh.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_MESH_DEBUG_PROXY)
+#define LOG_MODULE_NAME bt_mesh_gatt_client
+#include "common/log.h"
+
+#include "mesh.h"
+#include "adv.h"
+#include "net.h"
+#include "rpl.h"
+#include "transport.h"
+#include "host/ecc.h"
+#include "prov.h"
+#include "beacon.h"
+#include "foundation.h"
+#include "access.h"
+#include "proxy.h"
+#include "proxy_msg.h"
+#include "proxy_cli.h"
+#include "gatt_cli.h"
+#include "pb_gatt_cli.h"
+
+static struct bt_mesh_gatt_server {
+	uint16_t svc_start_handle;
+	uint16_t data_in_handle;
+	const struct bt_mesh_gatt_cli *gatt;
+
+	union {
+		void *user_data;
+		struct bt_gatt_discover_params discover;
+		struct bt_gatt_subscribe_params subscribe;
+	};
+} servers[CONFIG_BT_MAX_CONN];
+
+static struct bt_mesh_gatt_server *get_server(struct bt_conn *conn)
+{
+	return &servers[bt_conn_index(conn)];
+}
+
+static uint8_t notify_func(struct bt_conn *conn,
+			   struct bt_gatt_subscribe_params *params,
+			   const void *data, uint16_t length)
+{
+	const uint8_t *val = data;
+
+	if (!data) {
+		BT_WARN("[UNSUBSCRIBED]");
+		params->value_handle = 0U;
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (length < 1) {
+		BT_WARN("Too small Proxy PDU");
+		return BT_GATT_ITER_STOP;
+	}
+
+	(void)bt_mesh_proxy_msg_recv(conn, val, length);
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t discover_func(struct bt_conn *conn,
+			     const struct bt_gatt_attr *attr,
+			     struct bt_gatt_discover_params *params)
+{
+	int err;
+	struct bt_mesh_gatt_server *server = get_server(conn);
+
+	if (!attr) {
+		BT_DBG("GATT Services Discover complete");
+		(void)memset(params, 0, sizeof(*params));
```

Comment:
`param` was also NULL when I tried to provision a device.

Reply:
I don't understand what null means here?
Are prarms null pointers?
When I traverse all callbacks, it is impossible for params to be a null pointer.

Reply:
I can check, but I'm saying what I got when compiled this branch for `nrf52dk_nrf52832` board and tried to provision over pb-gatt.

Reply:
This is still an issue.
```
uart:~$ mesh provision-gatt deadbeef 0 2 0
Provisioning link opened on PB-GATT
[00:00:34.535,888] <dbg> bt_mesh_pb_gatt: prov_link_open: uuid deadbeef000000000000000000000000
[00:00:34.578,643] <dbg> bt_mesh_gatt_client: bt_mesh_gatt_cli_connect: Try to connect services
[00:00:34.685,363] <dbg> bt_mesh_gatt_client: gatt_connected: conn 0x20001ee8 err 0x00
[00:00:34.737,396] <dbg> bt_mesh_gatt_client: discover_func: [ATTRIBUTE UUID 0x2800] handle 12
[00:00:34.836,853] <dbg> bt_mesh_gatt_client: discover_func: [ATTRIBUTE UUID 0x2803] handle 13
[00:00:34.936,859] <dbg> bt_mesh_gatt_client: discover_func: [ATTRIBUTE UUID 0x2803] handle 15
[00:00:35.036,773] <dbg> bt_mesh_gatt_client: discover_func: [ATTRIBUTE UUID 0x2902] handle 17
[00:00:35.036,834] <dbg> bt_mesh_gatt_client: discover_func: [SUBSCRIBED]
[00:00:35.036,865] <dbg> bt_mesh_pb_gatt: bt_mesh_pb_gatt_start: conn 0x20001ee8
[00:00:35.041,625] <dbg> bt_mesh_gatt_client: bt_mesh_gatt_send: 3 bytes: 030000
[00:00:35.136,718] <dbg> bt_mesh_gatt_client: discover_func: GATT Services Discover complete
[00:00:35.136,749] <err> os: ***** MPU FAULT *****
[00:00:35.136,779] <err> os:   Data Access Violation
[00:00:35.136,779] <err> os:   MMFAR Address: 0x0
[00:00:35.136,810] <err> os: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000014
[00:00:35.136,810] <err> os: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x00023ba7
[00:00:35.136,810] <err> os:  xpsr:  0x81000000
[00:00:35.136,840] <err> os: Faulting instruction address (r15/pc): 0x0003f738
[00:00:35.136,871] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:35.136,901] <err> os: Current thread: 0x20002bd8 (BT RX)
[00:00:35.962,005] <err> os: Halting system
```

```
$ ~/nordic/gcc-arm-none-eabi-9-2019-q4-major/bin/arm-none-eabi-addr2line -e build/zephyr/zephyr.elf -a 0x0003f738
0x0003f738
/home/pavel/nordic/ncs/git-repos/zephyrproject/zephyr/lib/libc/minimal/source/string/string.c:390
$ ~/nordic/gcc-arm-none-eabi-9-2019-q4-major/bin/arm-none-eabi-addr2line -e build/zephyr/zephyr.elf -a 0x00023ba7
0x00023ba7
/home/pavel/nordic/ncs/git-repos/zephyrproject/zephyr/subsys/bluetooth/mesh/gatt_cli.c:143
```

Reply:
Please try again.

Reply:
This seems to be working now, thanks!

Reply:
Great